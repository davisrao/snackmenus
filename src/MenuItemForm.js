import React, { useState } from "react";
import "./MenuItemForm.css"
import ErrorMessage from "./ErrorMessage";
import SuccessMessage from "./SuccessMessage";

/**
 * Form for adding a new menu item to the site 
 * props:
 * - addMenuItem (fxn on handle submit that lives on parent)
 * 
 * state:
 * - formData
 * - addedItemSuccess (tracks validation success/fail)
 * - formErrors (tracks errors in form)
 * 
 * context:
 * - none
 * 
 * App => MenuItemForm =>{errorMessage, successMessage}
 */

function MenuItemForm({ addMenuItem }) {
    const initialFormData = {
        type: "drinks",
        name: "",
        description: "",
        recipe: "",
        serve: ""
    }

    const [formData, setFormData] = useState(initialFormData);
    const [addedItemSuccess, setaddedItemSuccess] = useState(false);
    const [formErrors, setFormErrors] = useState(null)
    console.log("* Menu Item form", addMenuItem, formData)


    /** function to check form data. returns nothing if form data correct
     * throws error if form data invalid
     */
    function validateFormData(formData) {
        for (let input in formData) {
            if (!formData[input]) {
                throw Error
            }
        }
    }

    /** Update form input. */
    function handleChange(evt) {
        const { name, value } = evt.target
        setFormData(oldData => ({
            ...oldData,
            [name]: value
        }))
    }

    /** Validate input, call parent function and clear form. 
     * Returns error message on improper validation.
    */
    async function handleSubmit(evt) {
        console.log()
        evt.preventDefault();
        try {
            validateFormData(formData)
            await addMenuItem(formData, formData.type);
            setFormData(initialFormData);
            setFormErrors(false)
            setaddedItemSuccess(true);
        } catch (err) {
            console.log(err);
            if (err.message === undefined) {
                setFormErrors("Form submission error. Ensure all fields are completed and try again!")
            } else {
                setFormErrors(`Form error. Ensure all fields are completed. 
                                If you still have trouble, check the menu to see i
                                f your item already exists.`)
            }
        }
    }

    return (
        <div className="">
            <form className="MenuItemForm" onSubmit={handleSubmit}>
                {addedItemSuccess && <SuccessMessage message={`Your submission was added successfully.`} />}
                {formErrors && <ErrorMessage message={formErrors} />}
                <div className="form-group">
                    <label htmlFor="type"
                        className="d-inline-flex">Select Type:&nbsp;&nbsp;
                    </label>
                    <select id="MenuItemForm-type"
                        name="type"
                        value={formData.type}
                        onChange={handleChange}
                        className="form-control form-control-sm d-inline-flex"
                    >
                        <option value={"drinks"}>Drink</option>
                        <option value={"snacks"}>Snack</option>
                    </select>
                </div>

                <div className="form-group">
                    <label htmlFor="name"
                        className="d-inline-flex">Name:&nbsp;&nbsp;
                    </label>
                    <input
                        id="MenuItemForm-name"
                        name="name"
                        className="form-control"
                        placeholder="Name"
                        onChange={handleChange}
                        value={formData.name}
                        aria-label="Name"
                    />
                </div>
                <div className="form-group">
                    <label htmlFor="description"
                        className="d-inline-flex">Description:&nbsp;&nbsp;
                    </label>
                    <input
                        id="MenuItemForm-description"
                        name="description"
                        className="form-control"
                        placeholder="Description"
                        onChange={handleChange}
                        value={formData.description}
                        aria-label="Description"
                    />
                </div>
                <div className="form-group">
                    <label htmlFor="recipe"
                        className="d-inline-flex">Recipe:&nbsp;&nbsp;
                    </label>
                    <textarea
                        id="MenuItemForm-recipe"
                        name="recipe"
                        className="form-control"
                        placeholder="Recipe"
                        onChange={handleChange}
                        value={formData.recipe}
                        aria-label="Recipe"
                    />
                </div>
                <div className="form-group">
                    <label htmlFor="serve"
                        className="d-inline-flex">Serving Instructions:&nbsp;&nbsp;
                    </label>
                    <textarea
                        id="MenuItemForm-serve"
                        name="serve"
                        className="form-control"
                        placeholder="Serve"
                        onChange={handleChange}
                        value={formData.serve}
                        aria-label="Serve"
                    />
                </div>

                <div className="form-group d-flex justify-content-between">
                    <button className="btn-primary rig btn btn-sm MenuItemForm-addBtn">
                        Add Item!
                    </button>
                </div>

            </form>
        </div>
    );
}

export default MenuItemForm;
