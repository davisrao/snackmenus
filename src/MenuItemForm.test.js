import React from "react";
import { render } from "@testing-library/react";
import MenuItemForm from "./MenuItemForm";

//run tests with npm test in terminal

describe("MenuItemForm Test", function () {
    it("matches snapshot", function () {
        const result = render(
            <MenuItemForm addMenuItem={() => null} />
        );
        expect(result).toMatchSnapshot();

    });
    it("renders properly with expected text", function () {
        const result = render(
            <MenuItemForm addMenuItem={() => null} />
        );
        console.log(result)
        expect(result.queryByText("Serving Instructions:")).toBeInTheDocument();
    });

});