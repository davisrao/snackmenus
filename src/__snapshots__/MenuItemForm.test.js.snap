// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItemForm Test matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class=""
      >
        <form
          class="MenuItemForm"
        >
          <div
            class="form-group"
          >
            <label
              class="d-inline-flex"
              for="type"
            >
              Select Type:  
            </label>
            <select
              class="form-control form-control-sm d-inline-flex"
              id="MenuItemForm-type"
              name="type"
            >
              <option
                value="drinks"
              >
                Drink
              </option>
              <option
                value="snacks"
              >
                Snack
              </option>
            </select>
          </div>
          <div
            class="form-group"
          >
            <label
              class="d-inline-flex"
              for="name"
            >
              Name:  
            </label>
            <input
              aria-label="Name"
              class="form-control"
              id="MenuItemForm-name"
              name="name"
              placeholder="Name"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class="d-inline-flex"
              for="description"
            >
              Description:  
            </label>
            <input
              aria-label="Description"
              class="form-control"
              id="MenuItemForm-description"
              name="description"
              placeholder="Description"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class="d-inline-flex"
              for="recipe"
            >
              Recipe:  
            </label>
            <textarea
              aria-label="Recipe"
              class="form-control"
              id="MenuItemForm-recipe"
              name="recipe"
              placeholder="Recipe"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class="d-inline-flex"
              for="serve"
            >
              Serving Instructions:  
            </label>
            <textarea
              aria-label="Serve"
              class="form-control"
              id="MenuItemForm-serve"
              name="serve"
              placeholder="Serve"
            />
          </div>
          <div
            class="form-group d-flex justify-content-between"
          >
            <button
              class="btn-primary rig btn btn-sm MenuItemForm-addBtn"
            >
              Add Item!
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class=""
    >
      <form
        class="MenuItemForm"
      >
        <div
          class="form-group"
        >
          <label
            class="d-inline-flex"
            for="type"
          >
            Select Type:  
          </label>
          <select
            class="form-control form-control-sm d-inline-flex"
            id="MenuItemForm-type"
            name="type"
          >
            <option
              value="drinks"
            >
              Drink
            </option>
            <option
              value="snacks"
            >
              Snack
            </option>
          </select>
        </div>
        <div
          class="form-group"
        >
          <label
            class="d-inline-flex"
            for="name"
          >
            Name:  
          </label>
          <input
            aria-label="Name"
            class="form-control"
            id="MenuItemForm-name"
            name="name"
            placeholder="Name"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class="d-inline-flex"
            for="description"
          >
            Description:  
          </label>
          <input
            aria-label="Description"
            class="form-control"
            id="MenuItemForm-description"
            name="description"
            placeholder="Description"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class="d-inline-flex"
            for="recipe"
          >
            Recipe:  
          </label>
          <textarea
            aria-label="Recipe"
            class="form-control"
            id="MenuItemForm-recipe"
            name="recipe"
            placeholder="Recipe"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class="d-inline-flex"
            for="serve"
          >
            Serving Instructions:  
          </label>
          <textarea
            aria-label="Serve"
            class="form-control"
            id="MenuItemForm-serve"
            name="serve"
            placeholder="Serve"
          />
        </div>
        <div
          class="form-group d-flex justify-content-between"
        >
          <button
            class="btn-primary rig btn btn-sm MenuItemForm-addBtn"
          >
            Add Item!
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
